runtime/parachains/src/paras.rs

```diff
@@ -408,40 +555,128 @@ impl<T: Config> Module<T> {
 		});
 
 		if dup {
-			let weight = T::DbWeight::get().reads_writes(1, 0);
+			weight = weight.saturating_add(T::DbWeight::get().reads(1));
 			return weight;
 		}
+		weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
 
+		if genesis.parachain {
+			ParaLifecycles::insert(&id, ParaLifecycle::OnboardingAsParachain);
+		} else {
+			ParaLifecycles::insert(&id, ParaLifecycle::OnboardingAsParathread);
+		}
 		UpcomingParasGenesis::insert(&id, &genesis);
+		weight = weight.saturating_add(T::DbWeight::get().writes(2));
 
-		T::DbWeight::get().reads_writes(1, 2)
+		weight
 	}
 
 	/// Schedule a para to be cleaned up at the start of the next session.
+	///
+	/// Noop if para is transitioning or already outgoing.
```

Comment:
Can a chain avoid being cleaned up (e.g. by the auctions system) by sneakily triggering a downgrade right before the cleanup is scheduled? As their `ParaLifecycle` would be `Downgrading`.

I think that a) they can, but b) it's not a big deal, since we expect that the auctions system would trigger upgrades and downgrades on behalf of paras. Although swaps could be used to get around that:
  * Para A's slot is almost up
  * Para A and Para B decide to swap
  * A = Downgrading, B = Upgrading
  * Auctions triggers downgrade of A, which is already downgrading so no-op
  * B stays as a parachain

As long as we intermediate swaps through the auctions module, this also can be prevented

Reply:
I think this is a good point, i see no reason not to allow clean up of a parachain that may be transitioning.

It would be the more "safe" thing to do

Reply:
Yeah, these kinds of issues would be exactly why explicit lifecycles are needed. I think as long as the downstream pallets call the lifecycle check, we should be free of issues, but I am open to making this function more on the "safe side" and have it also switch a transitioning chain to outgoing.